### PR TITLE
HDFDumpLayer: Add batch dim to "extra" inputs

### PR DESCRIPTION
**Pull request by Zettelkasten**
I had `extra` inputs of a HDF dump layer which did not have any batch dim.
Here I change it so that all `extra` inputs without batch dim get an unbroadcasted batch dim of the main input.

### CODE REVIEW

**Comment by patrick-wilken** [comment]
Yep, this is necessary.  Related to #331. Also here, we are limited by this `num_outputs = [dim, num_dims]` shape specification used by `Dataset` and also within the HDF file (`targets/size`). If we could specify `"batch_dim": None` etc. in the HDF file, we could implement it in a way that does not require unbroadcasting and therefore wasting space.

File: returnn/tf/layers/basic.py [hunk 1]
```diff
@@ -7677,7 +7677,11 @@ def __init__(self, filename, extra=None, dump_whole_batches=False, labels=None,
     self.filename = filename
     if extra is None:
       extra = {}
-    extra = {key: layer.output.copy_as_batch_spatial_major() for (key, layer) in extra.items()}
+    extra = {key: layer.output for (key, layer) in extra.items()}
+    extra = {
+      key: output.copy_add_batch_dim(0, batch=data.batch)
+      if not output.have_batch_axis() else output for (key, output) in extra.items()}
```
Comment (albertz) [inline returnn/tf/layers/basic.py:7683]:
```suggestion
      if not output.have_batch_axis() else output
      for (key, output) in extra.items()}
```

**Comment by albertz** [comment]
> If we could specify `"batch_dim": None` etc. in the HDF file, we could implement it in a way that does not require unbroadcasting and therefore saves space.

No. Our dataset concept has one entry per sequence. If you have some layer without batch dim, it means the same value/tensor will be used (broadcasted) for all sequences in the batch. What do you think the dataset (no matter if HDF or whatever else) should save per sequence (seq idx)?

> Related to #331. Also here, we are limited by this `num_outputs = [dim, num_dims]` shape specification used by `Dataset` and also within the HDF file (`targets/size`).

No, this is not really related (because see above). And also, I think I already mentioned, if `num_outputs` of the `Dataset` is really a problem, this is kind of trivial to fix / change / extend. We should not keep repeating that it is a problem if it is so. It probably would have been simpler to already fix this. But anyway, this is unrelated here.

**Review by albertz — APPROVED** [review]

**Comment by patrick-wilken** [comment]
A `Dataset` returns data per sequence and cannot provide a constant, I know. But I was talking about the HDF file structure. Per data key we store one big 1-dim array of values. What it represents is inferred from "metadata" (sequence lengths and the `size` field). I'm just saying it would be conceivable to add the information about a missing batch dim in the file and then interpret the whole data array for a given key as a constant sequence, and thus return it for each sequence index.

**Pull request merged**